[WIP] Compilers without support for flat float arrays

xclerc opened the pull request:
(This comment is *very* long, in an attempt to cover/discuss all cases.)

There seems to be various problems regarding the interaction between
distributions built with and without support for flat float arrays (FFA below).
These problems have a common root that is the production / expectation
regarding blocks with tag 254:

- a program/compiler with support for FFA will produce values with tag
  254 for the `float array`, `floatarray` and all-float record types;
- a program/compiler without support for FFA will expect values with tag
  254 only for `floatarray` and all-float record types, and will expect values
  with tag 0 for the `float array` type.

A program expecting a block with tag 0 but actually getting a block with tag
254 is more than likely to segfault at some point.

The current PR addresses various occurrences of this issue.


### Mixing modules

It is possible to produce a program that segfaults by compiling a module
with the following signature with a compiler with support for FFA:

    val f : ... -> float array

and then use it from a module compiled without support for FFA:

    let () = Array.iter print_float (Aux.f ...)

This can be fixed by adding a new flag to `cmi` files (akin to the flag for
unsafe strings) in order to avoid mixing modules. See commit `92d483ab`.

(For this to work properly, it needs the bytecode compiler in the `boot/`
directory to be compiled without support for flat float array. This is checked
by the travis script. See commits `8444618e` and `b03f3d84`.)


### Bytecode executables

A bytecode executable produced by a compiler with support for FFA can
segfault if executed with an `ocamlrun` VM compiled without support for FFA
if it contains a `MAKEFLOATBLOCK` instruction generated for a value of type
`float array`. For instance:

    let a = [| 0.; 1.; 2. |]
    let () = Array.iter print_float a

(Note that if this snippet is added to the code of the compiler, *e.g.* in
`driver/main.ml`, a segfault will occur while trying to build a distribution
configured with `-no-flat-float-array`.)

This is currently documented on the manual page for `ocamlrun`. See commit
`ec17a4d1`.

Another option would be to bump the magic number for bytecode executables,
and store the information about FFA support, refusing to execute a file if a
segfault might occur.

*Actual* fixes are also possible, depending on whether it is deemed important
to provide backward compatibility. For instance, some key array-related
primitives might keep their original semantics (*i.e.* include a dynamic check
for the kind of array) only for bytecode compilation. It is trivial for get/set/length
primitives but is slightly more complex for blit/gather (because these ones
currently assume that the various arrays involved in a given operation have the
same layout).


### Marshaling in user code

If a program compiled with support for FFA marshals a value of type `float array`
then unmarshaled by a program compiled without support for FFA, array uses
can segfault.

This is documented in the ocamldoc comment of the `Marshal` module.
See commit `9e5e2015`.


### Marshaling in compiler code, dynamic linking

While this is not currently the case, a compiler might marshal `float array`
values to `cm`*K* files. Then unmarshaling such values might cause segfaults.

This can be avoided by adding a boolean indicating whether the file might
contain flat arrays to all `cm`*K* files. The boolean is also used to ensure
that a program compiled without support for FFA does not try to dynamically
load files compiled with support for FFA. See commit `a817ad8c`.


### Miscellaneous

- Added a `Sys.flat_float_array` constant to give access to the information
  about FFA support (note: could be determined through user-land code *e.g.*
  `Obj.(tag @@ repr [| 0. |] = double_array_tag)`) - see commit `2b6d4cb9`;
- Updated sources in `tools/` - see commit `89680d8d`.


### TODO

This PR is still a work in progress; pending:

- (possibly) rewrite some changes above as they are are partially redundant;
- update chapter 20 ("Interfacing C with OCaml") of the manual;
- make sure that the various CI instances build variants without FFA support;
- create opam switches without FFA support to encourage testing.

xclerc committed:
Add a "flat float array" flag to cmi files.

xclerc committed:
Make Travis check that the bootstrap bytecode compiler has been be compiled without support for flat float arrays.

xclerc committed:
Document that an `ocamlrun` built without support for flat float array cannot be used to execute bytecode files compiled with such support.

xclerc committed:
Document that it is unsafe to send a `float array` value from a program with support for flat float arrays to a program without such support.

xclerc committed:
Add "flat float arrays" tags to `cm{i,o,x,xs}` to avoid [un]marshalling issues.

xclerc committed:
Define and use `Sys.flat_float_array`.

xclerc committed:
Update sources in `tools/` directory.

xclerc committed:
Miscellaneous.

xclerc committed:
Fix space in error message.

xclerc committed:
Fix analysis of "taken" match branches (flambda).

xclerc committed:
Avoid the creation of value of type `float array` (temporary workaround).

xclerc commented:
The changes above are not sufficient, as shown by the Travis failure
of the flambda variant. Indeed, an flambda compiler built using a
boostraping compiler without FFA support but run with an `ocamlrun`
with FFA support will fail, because the flambda compiler builds value
of type `float array`.

One solution is to avoid the construction of value with type `float array`,
(see commit `c668190e`) but it is brittle at best, because such values are
intermediary results (*e.g.* `Array.map (fun x -> Some x) (Array.of_list fs)`
with `fs` a list of `float` value in file `asmcomp/export_info.ml`).

Other solutions include:
- having the primitive retain their "old" semantics when in bytecode mode;
- restore the "old" semantics of array primitives and introduce new
  primitives, to be used ` in 4.07+.

damiendoligez commented:
I have a few remarks:

* It's nice to detect and prevent the mixing of files from differently-configured compilers.
* Cross-compatibility is not a goal, performance is. If you reintroduce run-time checks in the primitives, that defeats the whole purpose of the configure option. (re: your "other solutions")
* The only place where I'm interested in cross-compatibility is making the bootstrap easier or faster.
* I'm worried about the dissymetry of your code: you refuse FFA-marshalled values in no-FFA programs, but you accept the reverse. This is probably OK as long as you stay in OCaml, but any (user-defined) C primitive that deals with `float array` will crash or even return buggy results. I'll be more at ease if you just forbid mixing the two configurations altogether.

xclerc commented:
> I'm worried about the dissymetry of your code: you refuse FFA-marshalled values in no-FFA programs, but you accept the reverse. This is probably OK as long as you stay in OCaml, but any (user-defined) C primitive that deals with float array will crash or even return buggy results. I'll be more at ease if you just forbid mixing the two configurations altogether.

I understand your point, but there is currently only one bootstrap directory
that is used whatever the user pass to the configure script. Do you suggest
to have two bootstrap directories?

xclerc commented:
> any (user-defined) C primitive that deals with float array will crash or even return buggy results

I would expect the primitive to behave according to the value of `FLAT_FLOAT_ARRAY`
in `${PREFIX}/lib/ocaml/Makefile.config`.

damiendoligez commented:
> Do you suggest to have two bootstrap directories?

Ouch, no. I'll keep your not-completely-safe solution rather than duplicate the bootstrap binaries.

> I would expect the primitive to behave according to the value of `FLAT_FLOAT_ARRAY`
in `${PREFIX}/lib/ocaml/Makefile.config`.

Yes, so your FFA-configured program accepts a no-FFA marshalled value and crashes. I guess we'll have to live with that.

shindere committed:
config.h cleanup

Get rid of a few lines that are now useless

shindere committed:
Change the way ARCH_CODE32 is defined

Before this commit, this C preprocessor macro was defined in
byterun/caml/m.h by the configure script, but just on some architectures
and only in non-PIC mode.

This commit introduces the HAS_ARCH_CODE32 predicate which is inserted
in the m.h file when this is relevant, the ifdef block on PIC
being moved to config.h.

This is to prepare the switch to autoconf, since header files processed
by config.status are not allowed to contain ifdef blocks.

gasche committed:
Release highlights for 4.07.0 in a new 'News' file

nojb committed:
Use String.split_on_char

nojb committed:
Add Misc.split_path_contents

nojb committed:
Add support for OCAMLTOP_INCLUDE_PATH

nojb committed:
Doc in manual

nojb committed:
Changes

nojb committed:
Specify search order in doc, move to right place

nojb committed:
Amend doc

nojb committed:
doc: mention feature introduced in 4.08

nojb committed:
Fix regression: return empty list of directories when PATH-like env var is empty

nojb committed:
Changes

nojb committed:
Simplify

gasche committed:
add a minor script to turn ASCII references into Markdown links

gasche committed:
News: add 4.06.1

gasche committed:
News: add 4.06.0

gasche committed:
News: add 4.05.0

gasche committed:
check-typo: simplify implementation by deconstructing the built-in exceptions

gasche committed:
check-typo: make (--check-prune .git) work as expected

gasche committed:
rework the relation between .gitattributes and ./tools/check-typo

Before this patch, check-typo is directed by a single git attribute,
ocaml-typo, which is used as a key to set a value:

    ocaml-typo=long-line,missing-header

(the value here is `long-line,missing-header`, and the code splits the
comma later)

This model is very fragile because .gitattributes does not allow to
give attribute keys a collecting/aggregating semantic: each new
setting of the key removes the previous setting, instead of adding to
them. For example,

    testsuite/tests/**                      ocaml-typo=missing-header
    testsuite/tests/win-unicode/*.ml        ocaml-typo=utf8

and

    testsuite/tests/win-unicode/*.ml        ocaml-typo=utf8
    testsuite/tests/**                      ocaml-typo=missing-header

are not equivalent, and instead of using either one we would introduce
redundancy for robustness:

    testsuite/tests/**                      ocaml-typo=missing-header
    testsuite/tests/win-unicode/*.ml        ocaml-typo=missing-header,utf8

With this patch, we switch to a model where each ocaml-typo setting is
its own attribute, of the form `typo.<<attribute>>`. The lines above
would be written, in either order:

    testsuite/tests/**                      typo.missing-header
    testsuite/tests/win-unicode/*.ml        typo.utf8

Not only does this approach make our .gitattributes more robust, it
allows for a more fine-grained treatment of the "unused-prop"
marker. This was used as an attribute to say: don't make it in an
error if the given typo-rule is in fact respected (by default, opting
out of a typo-rule gives an error if the typo-rule is respected). But
because of the single-key nature of ocaml-typo, unused-prop would
range over all settings, not just one of them. For example

    emacs/caml.el ocaml-typo=long-line,unused-prop,missing-header

seems to suggest that 'unused-prop' only qualifies the 'long-line'
rule, but in fact it also ranges over 'missing-header'. In contrast,
with this patch, we write the following:

    emacs/caml.el typo.long-line=may typo.missing-header

the `=may` value setting is used to make an exception to a typo-rule
optional.
Interestingly, most .gitattributes lines worked without extra error
when I turned each unused-prop in a =may setting over the rule just
before, instead of all rules: our checking is now more precise than
before, better capturing the intent of the .gitattributes author.

As I had to rewrite parts of the check-typo code for this, I took the
opportunity to rename a couple variables speaking about SVN (now long
defunct) into more meaningful names:

- `$is_svn` => `$path_in_index`
- `$svnrules` => `$attr_rules`

gasche committed:
check-typo: make sure that 'prune' attributes on non-directories are honored

gasche committed:
check-typo: remove useless gitattributes

gasche committed:
Changes entry

xclerc commented:
Would it make sense to use the marshaling header (or magic number)
to detect incompatibilities to could lead to a segfault?

gasche committed:
rec_check.ml: fix a soundness bug with record override { foo with ... }

(The bug is already present in 4.06 and 4.07.)

gasche committed:
rec_check.ml: clarify the use of Guard for discarded values

gasche committed:
rec_check.ml refactoring: standard argument order for mode judgments

gasche committed:
rec_check.ml: clarify the pattern-matching rules

gasche committed:
rec_check.ml: write the term judgments in a more abstract style

We could factorize the binder judgments in a related way, but it would
require a different set of higher-order operator, and I thought that
the lower number of occurrences did not justify the extra complexity.

gasche committed:
Changes for the let-rec patchset + general commit-message explanation

This work was conducted by Alban Reynaud during an internship at INRIA
Saclay, with the help and under the supervision of Gabriel Scherer in
the Parsifal team.

A summary is included below. For more details, see

  https://github.com/Cemoixerestre/Internship-2018-ocaml-recursive-value

or the comments in the code itself, which has evolved a bit from
Alban's internship report above.

Summary:
--------

Rec_check is based on the idea of assigning "access modes" to variables

    m ::= Dereference (* full access and inspection of the value *)
        | Unguarded (* the value result is directly returned *)
        | Guard (* the value is stored under a data constructor *)
        | Delayed (* the value is not needed at definition time *)
        | Unused (* the value is not used at all *)

Before this patchset, the implementation (contributed by Jeremy Yallop)
was formulated as a type-checker manipulating "uses", which are mappings
from free variables to modes

    u ::= (x : m)*

and contexts Γ mapping variables to uses

    Γ ::= (x : u)*

The check relied on a judgment of the form

    Γ ⊢ e : u

where, in the typing algorithm, `Γ` and `e` are the inputs, and `u` is
the output — this could be written `-Γ ⊢ e : +u`.

After this patchset, the implementation uses a simpler notion of context
mapping variables to mods

    Γ ::= (x : m)*

and the judgment has the simpler structure

     Γ ⊢ e : m

but now `m` is an input to the judgment, indicating in which usage
context `e` is checked, and `Γ` is the output: `+Γ ⊢ e : -m`.

Other fixes:
------------

This patchset also fixes a soundness bug around `{ foo with ... }`
which is already present in 4.06 and 4.07.

Another soundness bug found during this work was submitted and fixed
independently as GPR #1915.

This work also led to the refactoring of GPR #1922, and the refactoring
of the testsuite in GPR #1937.

gasche committed:
rec_check: clarifications suggested by Armaël

trefis committed:
add some test

trefis committed:
printtyp: use scope instead of stamp for -short-paths

trefis committed:
printtyp: use ident map for naming context, not a stamp map

trefis committed:
ident: add an explicit scope field

- Ident.create now takes a scope as argument
- added Ident.create_var to use when the scope doesn't matter
- the current_time and the current_level are unrelated as of this
  commit. But one has to remember to bump the level when creating new
  scopes.

trefis committed:
ident: use a variant type instead of the flags field

trefis committed:
ident: split Local into Local and Scoped

Also rename [create] into [create_scoped] and [create_var] into
[create_local].

trefis committed:
ident: hide stamp related functions from the interface

The only remaining user was [Predef], which was bumping the counter by
999 to allow cmis to remain compatible if new predefs were defined.
This commit removes that use by making sure that every ident defined in
predef.ml is marked as predef and use a different stamp counter as user
defined idents. That way idents in cmi always start at 1, no matter how
many predefs there are.

Stamps on predef aren't strictly necessary: predefs names are unique.
However, predef idents comparison is just faster with stamps.

trefis committed:
testsuite: update

trefis committed:
Changes

dbuenzli committed:
Stdlib: add Bool module.

dbuenzli committed:
Stdlib: add Int module.

dbuenzli committed:
s/string_of_int/Int.to_string/g

dbuenzli committed:
Avoid opening Misc in asmcomp/selectgen.ml

This shadows the standard Stdlib module.

yallop committed:
explain why field assignments may dereference the new value

gasche committed:
renamed Unused into Ignore (suggested by Jeremy Yallop)

gasche committed:
clarifications suggested by Jeremy

gasche committed:
rec_check: clarify the mode ordering, add and use a ranking function

Suggested-by: Jeremy Yallop

gasche committed:
rec_check.ml: high-level documentation and reordering

We provide an ocamldoc-style module comment to explain the general
lines of the recursive check. This complements existing comments about
the term- and binding-judgments that are more implementation-oriented.

It was natural to reorder the code so that the "static or dynamic
size" code is apart from the "usage of recursive variables" code.

Finally, I also added copyright headers to reflect the now shared
authorship of the code. (Alban has signed a CLA.)

Octachron committed:
a type for unification traces

shindere committed:
Move config/Makefile to Makefile.config

In order to prepare the transition to autoconf, this commit moves the
configuration Makefile out of the config directory which will disappear
and gives it the name it will have once intstalled, namely Makefile.config.

hhugo committed:
cleanup lexer

rebase

trefis committed:
ident: remove create_hidden

trefis committed:
consistently make temporary local type idents local idents

trefis committed:
ident: fix unique name

trefis committed:
ident: fix doc

trefis committed:
ident: lowest_scope for globals, highest_scope for locals

trefis committed:
properly scope idents

- when including: the elements are rescoped to the current level (as
well as being given a fresh stamp, which was already the case)
- extension constructors idents cannot be local: they can be used as
types when the constructor's argument is an inlined record. They must be
given a scope
- in check_recmod_inclusion: create a new scope at each iteration
- when checking that type declarations inside recursive modules are well
founded, we now take a generic instance of the declaration (this is
reminiscent of what is done in Ctype.moregeneral)

gasche committed:
[minor] add rec_check.ml copyright info (as suggested by Damien)

nojb committed:
Add support for -args and -args0 to ocamlmklib

nojb committed:
Update Changes

gasche committed:
Merge pull request #1942 from gasche/alban-rec-check

New implementations of the static check for recursive definitions

damiendoligez committed:
Merge pull request #1683 from stedolan/marshal-custom-length

Make marshalled Custom_tag objects store their length.

trefis committed:
ident: better printing

gasche committed:
[minor] fix a macro bug from bd1a7c9b32e75

trefis committed:
create a new scope whenever we create new scoped idents

gasche committed:
Merge pull request #2045 from nojb/ocamlmklib_args

Add support for -args and -args0 to ocamlmklib

hhugo committed:
Sync ocamlex with the ocaml lexer

hhugo committed:
fix newline char

hhugo committed:
report error inside semantic action

hhugo committed:
improve errors

hhugo committed:
update error

hhugo committed:
CR

314eter committed:
Allow quoted strings and octal/unicode character literals in ocamllex actions (#1912)

shindere committed:
Introduce and use ROOTDIR in more makefiles

gasche committed:
Merge pull request #2049 from hhugo/lexer3

Fixes in lexers

trefis committed:
Unused rec: take local warning scope into account (#2050)

* unused rec: add test showing brokeness

* unused rec: allow turning the warning on locally

trefis committed:
call repr before checking level to know if a disambiguation is principal

shindere committed:
Break long lines introduced by 9046dc6219

trefis committed:
move refresh_signature out of Subst, add Env.enter_signature

garrigue committed:
Fix MPR#7818 by removing local aliases in functor argument types (#2051)

garrigue committed:
update output for pr7818.ml

cwfoo committed:
Fix typo in Scanf docs

'String.unescaped' -> 'Scanf.unescaped'

gasche committed:
Merge pull request #2054 from cwfoo/scanf-typo

Fix typo in Scanf docs

trefis committed:
bootstrap

trefis committed:
Merge pull request #1980 from trefis/scoped-idents

separate scope from stamp in idents

Octachron committed:
manual: fix misplaced end of itemize

gasche committed:
Merge pull request #2057 from Octachron/manual_fix_itemize

manual: fix misplaced end of itemize

mshinwell committed:
Add Lprologue (#2055)

shindere committed:
Remove spurious blank lines in .gitattributes

shindere committed:
Get rid of the standard_runtime configuration variable (#2066)

This configuration variable was formerly used by the -make_runtime and
-use_runtime option but this is no longer the case.

shindere committed:
Define OCAML_STDLIB_DIR in runtime/Makefile rather than in s.h (#2059)

mshinwell committed:
GPR#2056 (Backend_var)

trefis committed:
check_scope_escape: also handle type constructors escapes

trefis committed:
always associate a scope to a type (previously it was optional)

Previously, not having a scope meant the type was used in every context,
now we set the scope to "Btype.lowest_level" to mean the same thing.
The equivalence was made obvious by the recent changes to identifiers
scoping.

Wilfred committed:
Emacs: use symbol boundaries in regular expressions

\< is a word boundary, whereas \_< is a symbol boundary. Word
boundaries meant that we got confused by variables like in_foo or
end_foo.

This crashed imenu on files that contain variables with names of this
form.

Fixes
https://github.com/ocaml/tuareg/issues/190
https://caml.inria.fr/mantis/view.php?id=7850

mshinwell committed:
GPR#2076 (Targetint.print) (#2076)

samwgoldman committed:
Avoid page table lookup in Pervasives.compare with no-naked-pointers

samwgoldman committed:
Consolidate NO_NAKED_POINTER checks

Octachron committed:
Full explanation for unsafe cycles in recursive modules (#2058)

* Full explanation for unsafe cycles in rec modules

mshinwell committed:
Rename Mach.Ialloc record field from _words_ to _bytes_ and fix logic in a couple of places (#2074)

stedolan committed:
Merge pull request #2079 from samwgoldman/compare-no-foreign-pointers

Avoid page table lookup in Pervasives.compare with no-naked-pointers

nojb committed:
Fix windows ANSI locale: use CP_ACP instead of CP_THREAD_ACP (#2062)

* Fix windows ANSI locale: use CP_ACP instead of CP_THREAD_ACP

* Update Changes

mshinwell committed:
DWARF register numberings (#2080)

gasche committed:
add a release checklist (#1866)

* add a release checklist

Currently this list isn't publicly available, it sits in various
different versions on @damiendoligez's filesystems. He sent me a copy
when I took care of some of the recent releases. The present
presentation is a result of significant cleanups and changes to the
checklist -- in particular, some mistakes may have jumped in.

This is not a scripted process, it is very informal and it is likely
that there are some mistakes/omissions in the list. Yet, it sounds
better to have it somewhere in the source repository than not have it
around at all. It was certainly helpful to me, and it probably would
be to other release-help volunteers.

(The list is put in tools/ because there isn't a clearly better place
for it. This choice was suggested by Damien.)

samwgoldman committed:
Add Changes entry for GPR#2079 (#2089)

Octachron committed:
rewording for "constructor has no type" error

Commit:
Add a warning triggered on type declarations "type t = ()"

These are most likely a mistake for "type t = unit", but still valid, and lead
to quite confusing error messages afterwards because now `()` denotes two
different incompatible constructors.

Commit:
Changes entry

trefis committed:
Btype.set_{level,scope} only log actual changes

trefis committed:
Changes

trefis committed:
Revert "update locations for destructive substitutions"

This reverts commit f8a80702c069481b30349339d6822f92ba8d98f3.

trefis committed:
bootstrap

trefis committed:
refresh test output

trefis committed:
revert Changes

trefis committed:
add test for mpr 7852

gasche committed:
[minor] .mailmap update

trefis committed:
Merge pull request #2072 from trefis/always-scope

always associate a scope to a type (previously it was optional)

gasche committed:
Merge pull request #2091 from Armael/redefine_unit

Add a warning on type declarations "type t = ()"

Commit:
Add source highlighting for errors & warnings in batch mode

Octachron committed:
change entries collation

trefis committed:
Merge pull request #2092 from trefis/revert-1737

Fix MPR#7852 by reverting #1737

Commit:
Update tests

Commit:
Changes

dbuenzli committed:
Remove Bool.of_string.

dbuenzli committed:
Remove Int.of_string.

Commit:
Avoid losing information when using Debuginfo.from_location

emillon committed:
Fix link markup in Gc.finalise_last documentation

damiendoligez committed:
more fixes to release-checklist, from releasing 4.07.1

Commit:
Fix Unix.getaddrinfo when called on strings containing null bytes

nojb committed:
Merge pull request #1841 from nojb/toplevel_include_path_env_var

Add support for OCAMLTOP_INCLUDE_PATH environment variable

trefis committed:
Provide a way to build the bytecode compiler using Dune (#2093)

gasche committed:
Merge pull request #2078 from Wilfred/fix_imenu_word_boundaries

Emacs: use symbol boundaries in regular expressions

trefis committed:
correctly register dependencies

nojb committed:
Merge pull request #2099 from emillon/doc-markup-gc-finalise-last

Fix link markup in Gc.finalise_last documentation

gasche committed:
Merge pull request #2100 from Armael/getaddrinfo

Fix Unix.getaddrinfo when called on strings containing null bytes

damiendoligez commented:
I think the best solution is to change the magic number of mashaled files. You don't have room in the header for the FFA information.

damiendoligez commented on middle_end/allocated_const.mli:
```diff
@@ -25,7 +25,7 @@ type t =
   | Int64 of int64
   | Nativeint of nativeint
-  (* CR-someday mshinwell: consider using "float array" *)
+  (* CR-someday mshinwell: consider using "float array" (or "floatarray") *)
```
You should keep only the `"floatarray"` option : it will generate less trouble because it is compatible across configurations.

damiendoligez commented on otherlibs/dynlink/dynlink.ml:
```diff
@@ -332,7 +337,10 @@ let error_message = function
       "error loading shared library: " ^ reason
   | Inconsistent_implementation name ->
-      "implementation mismatch on " ^ name
+    "implementation mismatch on " ^ name
```
inconsistent indentation on this line

damiendoligez commented on stdlib/sys.mli:
```diff
@@ -147,6 +147,10 @@ val max_array_length : int
 
+val flat_float_array : bool
+(** Whether the compiler and runtime automagically flatten float arrays.
+    @since 4.07.0 *)
```
you need to update this tag :-(

damiendoligez commented on tools/objinfo.ml:
```diff
@@ -257,6 +257,13 @@ let read_dyn_header filename ic =
   with Failure _ | Sys_error _ -> None
 
 let dump_obj filename =
+  let check_float_float_arrays has_float_arrays =
```
This should be `check_flat_float_arrays`, right?

xclerc replied:
Indeed.

damiendoligez commented on tools/objinfo.ml:
```diff
@@ -266,12 +273,17 @@ let dump_obj filename =
     close_in ic;
+    if cu.cu_flat_float_arrays && not Config.flat_float_array then begin
+
+    end;
```
This strange conditional should be replaced with a call to `check_flat_float_arrays`.

damiendoligez commented on utils/config.mlp:
```diff
@@ -82,22 +82,22 @@ let flat_float_array = %%FLAT_FLOAT_ARRAY%%
-and cma_magic_number = "Caml1999A022"
+and cmi_magic_number = "Caml1999I023"
+and cmo_magic_number = "Caml1999O023"
+and cma_magic_number = "Caml1999A023"
```
Don't update the magic numbers in the PR, the policy is to bump them just before releasing the first beta of a major version.

damiendoligez reviewed (changes requested):
Looks good in general. Just a few requests.

Octachron committed:
improved messages after review

gasche committed:
Merge pull request #2094 from Octachron/no_type_rewording

Rewording for "constructor has no type" error

mshinwell committed:
Phantom let support in Clambda (#2060)

mshinwell committed:
Add [Proc.destroyed_at_reloadretaddr] (#2065)

Peaupote committed:
Change verbatim to caml_example in documentation (#2105)

* Add paragraph in manual's readme about signature option in caml_example env. Change verbatim env to caml_example.

mshinwell committed:
Phantom let support in Cmm (#2070)

gasche committed:
shebangs: use `/usr/bin/env bash` instead of `/bin/bash` for BSD compat

Currently our CI machine on FreeBSD fails to build with the following
error message:

```
bytecomp/generate_runtimedef.sh runtime/caml/fail.h runtime/primitives > bytecomp/runtimedef.ml
/bin/sh: bytecomp/generate_runtimedef.sh: not found
```

This PR converts all the files with a `/bin/bash` shebang to
`/usr/bin/env bash`. Some files use `/bin/sh` instead, and they are
unchanged. It may be the case that some 'bash' files could in fact run
using 'sh', but checking this requires more work and I decided to
follow the original author's intent.

(Personally I like using bash explicitly as it gives a consistent
programming environment that I can easily test on my machine; but this
should not be at the expense of portability.)

gasche committed:
downgrade two shell scripts from 'bash' to 'sh'

Portability review done by David Allsopp.

gasche committed:
Merge pull request #2107 from gasche/bash-env

shebangs: use `/usr/bin/env bash` instead of `/bin/bash` for BSD compat

gasche committed:
Merge pull request #1596 from Octachron/format_doc_max_indent

MPR#7720, format documentation: maximum indentation limit

gasche committed:
Changes entry moved to the right place

gasche committed:
mention 4.07.1 in the Changes

trefis committed:
make check_scope_escape behave more like update_level

Octachron committed:
Add context for unification errors

Commit:
update .depend

Octachron committed:
unification trace: reify object error earlier

gasche committed:
check-typo: silence minor warnings

gasche committed:
reorder the testsuite/tests/lib-scanf-2 test structure

The initial motivation is to have a `foo.reference` file
instead of `reference`.

dra27 committed:
Make typo.very-long-line => typo.long-line=may

typo.long-line can still be explicitly set in .gitattributes (and
typo.long-line typo.very-long-line=may will still issue unused if no
lines are more than 80 columns)

gasche committed:
address review comments by David and Damien

gasche committed:
tools/check-typo: add a a "--get-attrs <path>" mode for debug purposes

Octachron committed:
unification trace: update Changes

Octachron committed:
Merge pull request #2047 from Octachron/a_precise_type_for_unification_traces

A type for unification traces

shindere committed:
.mailmap: at line feed at end of file

trefis committed:
Merge pull request #2108 from trefis/check-constr-scope

check_scope_escape: also handle type constructors escapes

nojb committed:
Do not include duplicates in spelling suggestions

damiendoligez committed:
GPR#1910: rename .reference.* files

gasche committed:
Merge pull request #1910 from gasche/simpler-check-typo

Simpler check-typo

Commit:
Infer missing line numbers

Commit:
Update the tests

Commit:
Add option -error-style and environment variable OCAML_ERROR_STYLE

Commit:
Allow long lines in tests

Commit:
unified-options.etex: make check-typo happy

gasche committed:
Merge pull request #2096 from Armael/quote_location2

In error messages, print the source fragment responsible for the error

shindere committed:
Update test reference files

This is a follow-up to GPR#2096. Test reference files used when the
compiler was configured with -no-flat-float-array had not been updated.

shindere committed:
Remove useless lines from the Windows configuration makefiles

The OCaml definitions of topdir and wintopdir were actually
never used. It can be seen that the Unix configure script does not
define such variables.

shindere committed:
tools: move the definition of syslib from ocamlmklibconfig.ml to ocamlmklib.ml

shindere committed:
tools: move the definition of mklib from ocamlmklibconfig.ml to ocamlmklib.ml

yallop committed:
Disallow .~ in extended indexing operators. (#2106)

Disallow .~ in the lexer to preserve MetaOCaml compatibility.

gasche committed:
robustly prune manual files

shindere committed:
runtime: move the definition of INT64_LITERAL from m.h to config.h

shindere committed:
Build system: cosmetic fix

Under Unix, the LDFLAGS were passed twice when building the
programs in the runtime directory. This commit fixes this.

nojb committed:
Merge pull request #2011 from dbuenzli/int-support

Improve stdlib support for `int`

Chris00 committed:
Extend Stdlib.Float with more functions (#1794)

nojb committed:
Merge pull request #2010 from dbuenzli/bool-support

Improve stdlib support for `bool`

Octachron committed:
ocamldoc, manpage backend: unwrap module comment docstrings (#2114)

* unwrap non-code doctstring
* use .SS for ocamldoc titles
* update Changes

alainfrisch committed:
A public TODO list for cleaning up the typechecker implementation

alainfrisch committed:
Forgot note about code coverage

alainfrisch committed:
Take reviewer comments into account

alainfrisch committed:
Goals

alainfrisch committed:
Typo

shindere committed:
Remove the dl_needs_underscore feature (#2115)

This feature could not be enabled anyway

Octachron committed:
add makefile target for stdlib pdf

Octachron committed:
stdlib documentation: precedence table

Octachron committed:
doc: replace ordinal precedence with links

Octachron committed:
manual: #... operators are left associative

thvnx committed:
Support FMA operation (#1354)

Adds a fused multiply-add operation to the Float module.

The following changes are made:
- configure: check math.h for the C99 fma() operation.
- fma declarations in float.ml[i] (stdlib/).
- C fma() call or emulation in runtime/floats.c.
- dedicated tests in testsuite/tests/fma.

Octachron committed:
review: typo + synchronisation comment

Octachron committed:
make precedence appendix an appendix

gasche committed:
clarify Set.diff documentation (#2119)

fixes [MPR#7868](https://caml.inria.fr/mantis/view.php?id=7868)

thvnx committed:
Update fma tests for Cygwin32 (#2120)

This patch updates accepted fma test results for the Cygwin32
platform.

gasche committed:
News: add 4.07.1

gsg committed:
Fix #mod_use raising exception (MPR#7867)

gasche committed:
Merge pull request #2124 from gsg/fix-mod-use-exception

Fix #mod_use raising exception (MPR#7867)

gasche committed:
Merge pull request #1834 from gasche/4.07-change-highlights

Include release highlights in a 'News' file

gasche committed:
testsuite: convert typing-unboxed-types to expect-style tests

This will be very useful for an upcoming PR implementing a different
unboxing check, which change the result for some of the tests. Without
the present change, the diff in test results was unreadable, which
is problematic for development and for code review.

trefis committed:
test nondep

trefis committed:
typemod: fix uncaught Nondep_cannot_erase exception

trefis committed:
Merge pull request #2126 from trefis/nondep-fix

Fix uncaught Nondep_cannot_erase exception

gasche committed:
Merge pull request #2116 from alainfrisch/typechecker_todo

TODO list for cleaning up the typechecker implementation

gasche committed:
Merge pull request #2128 from gasche/unboxed-types-testsuite

testsuite: convert typing-unboxed-types to expect-style tests

nojb committed:
Update Changes

alainfrisch committed:
Fix printing of type variables with a quote on 2nd character

Type variables whose name has a single quote as the second character
(such as ' a' or ' a'b) need to be printed with a whitespace after the
initial quote symbol, in order not to be confused with a character
literal.

yallop committed:
Warn on literal patterns found anywhere in a constructor's arguments.

Octachron committed:
documentation: move operator table to a new page

Octachron committed:
documentation: no header for mld files

nojb committed:
Add test

yallop committed:
Generalized warn_on_literal_pattern: test cases for deep matching

Armael committed:
Merge pull request #2109 from nojb/spellchecking_disambiguation

Print full paths in spellchecking suggestions when needed for disambiguation

trefis committed:
Fix wrong register typing in float (and boxed int) unboxing (#2083)

alainfrisch committed:
Move variable printer to Pprintast

- The code responsible for printing Syntaxerr errors is moved to the
  Parse module (so that it can depend on the variable printer in
  Pprintast).

- Pprintast becomes a dependency for a few tools that link some
  compiler modules in an ad hoc way (they would better be implemented
  in terms of compiler-libs).

yallop committed:
Generalized warn_on_literal_pattern: Changes entry.

yallop committed:
Convert warning 52 tests to expect style.

yallop committed:
Update the manual for the generalized warning 52.

Octachron committed:
GPR2117 changes

Octachron committed:
unification trace: properly track universal variable renaming (#2135)

* unification trace: track univar renaming

This PR fixes an inconsistency introduced by #2047 : when an universal type variable was renamed in Printtyp, the explanation part of the error message kept the original name leading to potentially confusing error messages.

This PR fixes this inconsistency by keeping the whole type expression in the explanation part of the unification message instead of just its original name (and add a test for this behavior).

dbuenzli committed:
Stdlib: add Fun module. (#2129)

* Stdlib: add Fun module.

* Stdlib: rename Bool.negate to Fun.negate.

Octachron committed:
pdf manual:fix refs, compilerlibs/ocaml_operators

damiendoligez committed:
Add caml_alloc_custom_mem (#1738)

* add caml_alloc_custom_mem and corresponding GC parameters
* fix a bug in tests/misc/ephetest2.ml

Octachron committed:
ocamldoc makefile: clean pdf and texi docs

Octachron committed:
documentation: fix a warning in latex mode

Octachron committed:
Merge pull request #2117 from Octachron/stdlib_precedence_table_3

documentation: precedence table for the standard library.

gasche committed:
Merge pull request #2133 from yallop/warn-on-literal-full

Warn on literal patterns found anywhere in a constructor's arguments.

nojb committed:
Merge pull request #2012 from dbuenzli/deprecate-string_of_int

s/string_of_int/Int.to_string/g

alainfrisch committed:
Filename.chop_suffix_opt (#2125)

xclerc committed:
Address review comments.

xclerc committed:
Address review comments.

xclerc committed:
Forgot to restore some magic numbers.

xclerc committed:
Forgot to restore some magic numbers.

xclerc committed:
Forgot a magic number.

xclerc committed:
Forgot a magic number.

xclerc committed:
Rebase mistake.

xclerc committed:
Merge

damiendoligez commented:
> Would it make sense to use the marshaling header (or magic number)
to detect incompatibilities to could lead to a segfault?

Yes, it would!

gasche commented:
What is the status of this PR? Is there still interest in working on this? (There was an issue with the rebase apparaently, which should be re-done to keep a clean patchset.)

damiendoligez commented:
@xclerc feel free to reopen if/when you want to resume work on this.
